roxie/ccd/ccdfile.cpp

```diff
@@ -1270,6 +1619,222 @@ class CRoxieFileCache : implements IRoxieFileCache, implements ICopyFileProgress
+                {
+                    keyIndex.setown(createKeyIndex(fileName, localFile->getCrc(), *localFile.get(), fileIdx, false, false));  // MORE - we don't know if it's a TLK, but hopefully it doesn't matter
+                    if (!keyIndex)
+                        keyFailed = true;
+                }
+                if (inNodeCache && keyIndex)
+                {
+                    // Round startOffset up to nearest multiple of index node size
+                    unsigned nodeSize = keyIndex->getNodeSize();
+                    startOffset = ((startOffset+nodeSize-1)/nodeSize)*nodeSize;
+                    do
+                    {
+                        bool loaded = keyIndex->prewarmPage(startOffset);
+                        if (!loaded)
+                            break;
+                        preloaded++;
+                        startOffset += nodeSize;
+                    }
+                    while (startOffset < endOffset);
+                }
+                else if (file_mmap)
+                {
+                    if (fileIdx != (unsigned) -1)
+                        noteRead(fileIdx, startOffset, (endOffset-1) - startOffset);  // Ensure pages we prewarm are recorded in our cache tracker
+                    do
+                    {
+                        if (startOffset >= (offset_t) file_stat.st_size)
+                            break;    // Let's not core if the file has changed size since we recorded the info...
+                        t += file_mmap[startOffset];  // NOTE - t reported below so it cannot be optimized out
+                        touched++;
+                        startOffset += os_page_size;
+                    }
+                    while (startOffset < endOffset);
+                }
+
+                if (*cacheInfo != ' ')
+                    break;
+                cacheInfo++;
+            }
+#ifndef _WIN32
+            if (file_mmap)
+                munmap(file_mmap, file_stat.st_size);
+#endif
+            if (fd != -1)
+                close(fd);
+            if (*cacheInfo != '\n')
+                break;
+            cacheInfo++;
+        }
+        assert(!*cacheInfo);
+        if (*cacheInfo)
+        {
+            DBGLOG("WARNING: Unrecognized cacheInfo format at %.20s", cacheInfo);
+        }
+        if (traceLevel)
+            DBGLOG("Touched %d pages, preloaded %d index nodes, result %d", touched, preloaded, t);  // We report t to make sure that compiler doesn't decide to optimize it away entirely
+    }
+
+    virtual void clearOsCache() override
+    {
+        if (activeCacheReportingBuffer)
+            activeCacheReportingBuffer->clear();
+    }
+
+    virtual void reportOsCache(StringBuffer &ret, unsigned channel) const override
+    {
+        if (activeCacheReportingBuffer)
+        {
+            Owned<CacheReportingBuffer> temp = new CacheReportingBuffer(*activeCacheReportingBuffer);
```

Comment:
I see a CS in getCacheInfo() but should it be before we create the temp ?  Could it change inbetween creating and reporting ?

Reply:
Yes, but that's ok.